### PR TITLE
Pattern matching with `KClass` and `KProperty`

### PR DESCRIPTION
**Pull request by serras**
This is a generalization of #3442, but using `kotlin.reflect` APIs instead of optics (and only the part in the standard library, so we actually do _not_ depend on the `kotlin.reflect` library). The result doesn't look so nice, but it can still be very useful.

```kotlin
val User.shownName: String get() = this.matchOrThrow {
  Person::class.of(Person::name.of(Name::firstName), Person::age.suchThat { it < 18 }) then { (fn, _) -> fn }
  Person::class.of(Person::name.of(Name::firstName, Name::lastName)) then { (fn, ln) -> "Sir/Madam $fn $ln" }
  Company::class.of(Company::name, Company::director.of(Name::lastName)) then { (nm, d) -> "$nm, att. $d" }
}
```

Implementation note: I started with even shorter code, without `.of` (`Person::name(Name::firstName)`). However, it turns out that this is reserved syntax, so I had to come up with a small name instead.

### CODE REVIEW

**Comment by nomisRev** [comment]
Sorry for the hold-up @serras. This is on my backlog for asap.

**Review by nomisRev — APPROVED** [review]
Couple of small nits, and questions.

Also, does arrow-match belong to arrow-optics if it doesn't depend on optics? Maybe we can consider putting it completely separate, or under core as well. Wherever you think this fits best.

File: arrow-libs/optics/arrow-match/src/commonMain/kotlin/arrow/match/Builders.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * to signal that it does not match the value.
+ */
+public typealias Matcher<S, A> = KProperty1<S, A>
+
+public expect fun <S, A> Matcher(
+  name: String,
+  get: (S) -> A
+): Matcher<S, A>
+
+public object DoesNotMatch: Throwable()
```
[the remark this answers is not in the document]
Comment (serras) [reply arrow-libs/optics/arrow-match/src/commonMain/kotlin/arrow/match/Builders.kt]: Changed to `expect`/`actual`

File: arrow-libs/optics/arrow-match/src/commonMain/kotlin/arrow/match/Builders.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+public fun <S> self(): Matcher<S, S> = Matcher("self") { it }
```
[the remark this answers is not in the document]
Comment (serras) [reply arrow-libs/optics/arrow-match/src/commonMain/kotlin/arrow/match/Builders.kt]: I've changes this to `identity`

File: arrow-libs/optics/arrow-match/src/commonMain/kotlin/arrow/match/Builders.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+public inline fun <S, reified A: S> it(): Matcher<S, A> =
```
[the remark this answers is not in the document]
Comment (serras) [reply arrow-libs/optics/arrow-match/src/commonMain/kotlin/arrow/match/Builders.kt]: This is not allowed, but I've changed it to `instanceOf` to align it with `arrow-optics` version

**Comment by serras** [comment]
> Also, does arrow-match belong to arrow-optics if it doesn't depend on optics?

I'm not really sure about where it should be. In my mind it goes in optics because it's about immutable data inspection, like optics.

Also, I fear we end up with a super-big `core` folder.

**Pull request merged**